File: scala/luhn/src/main/scala/Luhn.scala


object Luhn {
  def valid(str: String): Boolean = {


    val ltrs = str.filter(!_.isSpaceChar)

    if (ltrs.length < 2 || ltrs.exists(!_.isDigit)) {
      return false
    }

    ltrs.toCharArray.reverse.map { _.asDigit }.zipWithIndex.map { i =>
      var dbl = i._1
      if (i._2 % 2 == 1) dbl *= 2
      if (dbl > 9) dbl -= 9
      dbl
    }.sum % 10 == 0

  }

}

File: haskell/luhn/src/Luhn.hs
module Luhn (isValid) where
import Data.Char (isDigit, digitToInt)

isValid :: String -> Bool
isValid n | length (normalize n) < 2 = False
          | any (not . (\x -> x==' ' || isDigit x)) n = False
          | otherwise = mod (luhn $ normalize n) 10 == 0
  where
    normalize = map digitToInt . reverse . filter isDigit
    luhn (x:y:xs) = x + (if y<5 then 2*y else 2*y-9) + luhn xs
    luhn [x] = x
    luhn [] = 0
